aten/src/ATen/native/Unique.cpp

```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+    const bool equal_nan = false) {
+  const Tensor& input = self.contiguous();
+
+  int64_t numel = input.numel();
+  Tensor output = at::empty({0}, self.options());
+  Tensor inverse_indices = at::empty({0}, self.options().dtype(kLong));
+  Tensor counts = at::empty({0}, self.options().dtype(kLong));
+
+  if (numel == 0) {
+    if (return_inverse) {
+      inverse_indices.resize_(input.sizes());
+    }
+    return std::make_tuple(output, inverse_indices, counts);
+  }
+
+  // index of first unique in each consecutive section
+  // this is used to compute counts for parallelization purpose
+  Tensor unique_index = at::empty({0}, self.options().dtype(kLong));
+
+  // original behavior with unique on scalar tensor
+  // is to return a output size of ([1]), `flatten` here will do the job
+  auto input_flattened = input.flatten();
+
+  Tensor input_sorted, indices;
+  std::tie(input_sorted, indices) = input_flattened.sort();
+
+  scalar_t* input_sorted_data = input_sorted.data_ptr<scalar_t>();
+  int64_t* indices_data = indices.data_ptr<int64_t>();
+
+  // `mask` keeps track of whether it is the first unique
+  // in the sorted input sequence
+  Tensor mask = at::empty({numel}, self.options().dtype(kBool));
+  auto mask_acc = mask.accessor<bool, 1>();
+
+  int num_threads = at::get_num_threads();
+  std::vector<int64_t> unique_count_thread(num_threads, 0);
+  std::vector<int64_t> offset_thread(num_threads, 0);
+
+  // the first element is always true
+  mask_acc[0] = true;
+
+  // we can parallel on [1, numel) but we need to make sure it has
+  // the same parallel scope with the next loop
+  at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
+    if (begin == 0) { begin += 1; }
+    for (const auto i : c10::irange(begin, end)) {
+      mask_acc[i] = input_sorted_data[i] != input_sorted_data[i - 1];
+    }
+  });
+
+  // handle equal the NaN's
+  bool last_element_isnan = _isnan<scalar_t>(input_sorted_data[numel - 1]);
+  if (input.is_floating_point() && equal_nan && last_element_isnan) {
+    int64_t firstnan_index = numel;
+    while (_isnan<scalar_t>(input_sorted_data[firstnan_index - 1])) {
+      firstnan_index--;
+    }
+    mask_acc[firstnan_index] = true;
+    if (firstnan_index + 1 < numel) {
+      for (const auto i : c10::irange(firstnan_index + 1, numel)) {
+        mask_acc[i] = false;
+      }
+    }
+  }
+
+  // calculate unique count from each thread
+  at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
+    int tid = at::get_thread_num();
+    for (const auto i : c10::irange(begin, end)) {
+      if (mask_acc[i]) {
+        unique_count_thread[tid]++;
+      }
+    }
+  });
+
+  // calculate total count of unique and thread offset in output
+  int64_t unique_count = std::accumulate(unique_count_thread.begin(), unique_count_thread.end(), 0);
+  int64_t sum = 0;
+  for (const auto t : c10::irange(num_threads)) {
+    offset_thread[t] = sum;
+    sum += unique_count_thread[t];
+  }
```

Comment:
nit: seems like `sum` and `unique_count` in the end are the same? So, I guess, these two loops could be merged...

Reply:
Oh yes, thanks for the remainder !